cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringProgramEventPublisher.java

```diff
@@ -0,0 +1,232 @@
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+
+/**
+ * Reads program status messages from TMS and writes to them to peer specific topics.
+ * {@link TetheringAgentService} is responsible for reading status messages from the peer specific topics and
+ * sending updates to the respective peers.
+ */
+public class TetheringProgramEventPublisher extends AbstractRetryableScheduledService  {
+  private static final Logger LOG = LoggerFactory.getLogger(TetheringProgramEventPublisher.class);
+  private static final Gson GSON = ApplicationSpecificationAdapter.addTypeAdapters(new GsonBuilder())
+    .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
+    .registerTypeAdapter(ProgramOptions.class, new ProgramOptionsCodec())
+    .create();
+  private final TetheringStore store;
+  private final MessagingService messagingService;
+  private final MessageFetcher messageFetcher;
+  private final MessagePublisher messagePublisher;
+  private final ProgramRunRecordFetcher runRecordFetcher;
+  private final TransactionRunner transactionRunner;
+  private final String programUpdateTopic;
+  private final int programUpdateFetchSize;
+  private final String programStateTopicPrefix;
+  private final long pollInterval;
+
+  // Tracks id of last program status update message that was processed.
+  private String lastProgramUpdateMessageId;
+
+  @Inject
+  TetheringProgramEventPublisher(CConfiguration cConf, TetheringStore store, MessagingService messagingService,
+                                 ProgramRunRecordFetcher programRunRecordFetcher, TransactionRunner transactionRunner) {
+    super(RetryStrategies.fromConfiguration(cConf, "tethering.agent."));
+    this.store = store;
+    this.messagingService = messagingService;
+    this.messageFetcher = new MultiThreadMessagingContext(messagingService).getMessageFetcher();
+    this.messagePublisher = new MultiThreadMessagingContext(messagingService).getMessagePublisher();
+    this.runRecordFetcher = programRunRecordFetcher;
+    this.transactionRunner = transactionRunner;
+    this.programUpdateTopic = cConf.get(Constants.AppFabric.PROGRAM_STATUS_RECORD_EVENT_TOPIC);
+    this.programUpdateFetchSize = cConf.getInt(Constants.AppFabric.STATUS_EVENT_FETCH_SIZE);
+    this.programStateTopicPrefix = cConf.get(Constants.Tethering.PROGRAM_STATE_TOPIC_PREFIX);
+    // TetheringAgentService reads messages published by TetheringProgramEventPublisher, so use the same
+    // poll interval here
+    this.pollInterval = TimeUnit.SECONDS.toMillis(cConf.getLong(Constants.Tethering.CONNECTION_INTERVAL));
+  }
+
+  @Override
+  protected void doStartUp() {
+    TransactionRunners.run(transactionRunner, context -> {
+      AppMetadataStore appMetadataStore = AppMetadataStore.create(context);
+      lastProgramUpdateMessageId = appMetadataStore.retrieveSubscriberState(programUpdateTopic,
+                                                                            TetheringAgentService.SUBSCRIBER);
```

Comment:
this constant should be moved to this class

Reply:
done